### ISSUE TITLE
A phoneme of a child table is not used if that phoneme is not also defined in its parent table

### ISSUE THREAD

**Issue by BenTalagan** [espeak-ng/espeak-ng#679]
Hi, I'm not totally sure if it's a bug or not, but it's a bit obscure to me. Here is my use case. 

For transcription purpose, I'd like to rewrite my own phoneme tables for english, and I've stumbled on the following oddity.

I keep the original `en_rules` and `en_dict` files, but I'm rewriting a new english voice, called `en-test` (file `espeak-ng-data/lang/gmw/en-test`), with the following content : 

```
name Test English
language en-test  42
language en 42
phonemes en-test
tunes s1 c1 q1 e1
```

I'm declaring the new phoneme table like this (in the `phonemes` file) : 

`file : phonemes`
```
phonemetable en-test base1
include ph_english_test
```

Note that it does not inherit from the `en` table, but directly from `base1`.

Finally, I'm declaring the `ph_english_test` phoneme file (in `phsource/ph_english_test`). It's really simple, with only one phoneme (I've reduced the example to a minimal test case to illustrate my problem) : 

`file : phsource/ph_english_test`
```
phoneme a#
  vwl
  ipa £
endphoneme
```

Now, if I try to translate the following example, the `a#` phoneme is not used : 

```shell
> espeak-ng against -v en-test -X --ipa
Found: 'against' [a#gEnst]  
əɡˈənst
```

The strange thing is, if I add the phoneme to `base1` in the `phonemes` file, like this (note the different ipa character) : 

`file : phonemes`
```
phoneme a#
  vwl
  ipa $
endphoneme
```

then the phoneme is used, and it is taken from the **child** table :  

```shell
> espeak-ng against -v en-test -X --ipa
Found: 'against' [a#gEnst]  
ˈ£ɡənst
```

Finally, if I remove the `a#` phoneme from the child table, then the phoneme is still found and used, and it comes from `base1`, the **parent** table : 

```shell
> espeak-ng against -v en-test -X --ipa
Found: 'against' [a#gEnst]  
ˈ$ɡənst
```

To summarize, here are the four possible cases for the parent and child tables : 

Parent : not defined, Child : not defined => Not translated (pretty logical)
Parent : defined, Child : defined => translated, uses child table definition (pretty logical)
Parent : defined, Child : not defined => translated, uses parent table definition (pretty logical)
Parent : not defined, Child : defined => Not translated (seems odd)

Is this behavior really normal? I would have expect the `a#` phoneme to be working without having to define it in the parent table, just in the child table. Is there something I'm doing wrong or is it a bug?

**Comment by valdisvi** [comment]
Did you [rebuild entire project](https://github.com/espeak-ng/espeak-ng/blob/master/docs/building.md#building-1) or [recompiled phoneme tables](https://github.com/espeak-ng/espeak-ng/blob/master/docs/add_language.md#building-phonemes) before tests?

**Comment by BenTalagan** [comment]
Thank you @valdisvi  for your answer! Yes. I've also added the `ph_english_test` file to the list of phoneme files in `Makefile.am`. I always run `make && make install` before performing these tests. I've just checked again the full described behavior after you've answered, and I get the same results as yesterday.

**Comment by BenTalagan** [comment]
Ok, I think I've found the reason of the problem. However I'd be happy to have a confirmation.

I believe what I'm doing is illegal: I'm defining two different phoneme systems for the same dictionary :

`base > base1 > en`
`base > base1 > en-test`

It looks like a dictionary is precompiled against a phoneme system (en_dict for example is precompiled against the `en` phoneme system). By inheriting from `base1` for `en-test` directly but using the `en_dict` precompiled against the `en` system, `en-test` is not compliant with the precompiled dictionary (it is kind of 'detached' from it) and it can lead to obscure results.

I guess that, if I want to make things working, I should inherit from the `en` phoneme table and overload **every** phoneme, or alternatively, create a new language by duplicating en_rules, en_dict, en_emoji, and en_extra. Am I right ?

**Comment by valdisvi** [comment]
I think, you just need to extend en phonemes (`phonemetable en-test en`) and override only needed ones. As you have setting `phonemes en-test` in the language definition file, customized phonemes should be compiled into different `en-test_dict` file.

**Comment by BenTalagan** [comment]
Thanks for your answer! I think the problem was in the naming of the dict file(s). Since `en-test_dict` was not compiled at all, `espeak-ng` would default to `en` dict. Anyway, I'll do as you advise, I believe now it is a wiser approach.

**Comment by BenTalagan** [comment]
This was not a bug.
